wicker/core/datasets.py

```diff
@@ -89,26 +160,53 @@ def chunk_data_for_split(chunkable_data: List[Any], chunk_number: int = 500) ->
     return local_chunks
 
 
-def iterate_bucket_key_chunk_for_size(bucket_key_locs: List[Tuple[str, str]]) -> int:  # type: ignore
+def iterate_bucket_key_chunk_for_size(
+    input_tuple: Tuple[List[Tuple[str, str]], bool, Optional[storage.Client], Optional[storage.Bucket]]
+) -> int:  # type: ignore
     """Iterate on chunk of s3 files to get local length of bytes.
 
+    Input tuple has structure indicated below, contains the info required to get file and
+    optionally copy to gcloud. Structure is described below.
+
     Args:
-        bucket_key_locs: List of Tuple[str, str] containing the s3 bucket and key to check size.
+        bucket_key_locs List of Tuple[str, str]: containing the s3 bucket and key to check size.
+        copy_to_gcloud bool: bool specifying to copy to gcloud or not.
+        gcloud_client storage.Client: GCloud client for copying data or None if no copying
+        gcloud_bucket storage.Bucket: GCloud bucket object to copy data.
 
     Returns:
         int: total amount of bytes in the file chunk list
 
     """
+    bucket_key_locs, copy_to_gcloud, gcloud_client, gcloud_bucket = input_tuple
     local_len = 0
 
     # create the s3 resource locally and don't pass in. Boto3 docs state to do this in each thread
     # and not pass around.
     s3_resource = boto3.resource("s3")
+    lock = Lock()
     for bucket_key_loc in bucket_key_locs:
         bucket_loc, key_loc = bucket_key_loc
         # get the byte length for the object
         byte_length = s3_resource.Object(bucket_loc, key_loc).content_length
         local_len += byte_length
+
+        if copy_to_gcloud:
+
+            with lock:
+                config = get_config()
+                tmp_output_loc = config.gcloud_storage_config.local_gcloud_tmp_data_transfer_dir
+                gcloud_wicker_root_path = config.gcloud_storage_config.bucket_wicker_data_head_path
```

Comment:
@convexquad this is the locking I mentioned, not sure imo, I don't like each thread making a new one but I don't know enough about python locks